{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate list of stromal genes with negative association with TCGAov tumour cellularity (a.k.a. purity)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Script related to Consensus gene sets\n",
    "\n",
    "### Aim:\n",
    "- Generate list of stromal genes with negative correlation with TCGAov tumour purity, as TIMER has done for immune cells, in this case for stromal cells (endothelial and fibroblasts so far)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.6.8 :: Anaconda custom (64-bit)\r\n"
     ]
    }
   ],
   "source": [
    "!python --version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "home = \"~/git_repos/HGSOC_TME_Heterogeneity/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load ESTIMATE's stromal genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['HALLMARK_TNFA_SIGNALING_VIA_NFKB', 'HALLMARK_HYPOXIA',\n",
       "       'HALLMARK_CHOLESTEROL_HOMEOSTASIS', 'HALLMARK_MITOTIC_SPINDLE',\n",
       "       'HALLMARK_WNT_BETA_CATENIN_SIGNALING',\n",
       "       'HALLMARK_TGF_BETA_SIGNALING', 'HALLMARK_IL6_JAK_STAT3_SIGNALING',\n",
       "       'HALLMARK_DNA_REPAIR', 'HALLMARK_G2M_CHECKPOINT',\n",
       "       'HALLMARK_APOPTOSIS', 'HALLMARK_NOTCH_SIGNALING',\n",
       "       'HALLMARK_ADIPOGENESIS', 'HALLMARK_ESTROGEN_RESPONSE_EARLY',\n",
       "       'HALLMARK_ESTROGEN_RESPONSE_LATE', 'HALLMARK_ANDROGEN_RESPONSE',\n",
       "       'HALLMARK_MYOGENESIS', 'HALLMARK_PROTEIN_SECRETION',\n",
       "       'HALLMARK_INTERFERON_ALPHA_RESPONSE',\n",
       "       'HALLMARK_INTERFERON_GAMMA_RESPONSE', 'HALLMARK_APICAL_JUNCTION',\n",
       "       'HALLMARK_APICAL_SURFACE', 'HALLMARK_HEDGEHOG_SIGNALING',\n",
       "       'HALLMARK_COMPLEMENT', 'HALLMARK_UNFOLDED_PROTEIN_RESPONSE',\n",
       "       'HALLMARK_PI3K_AKT_MTOR_SIGNALING', 'HALLMARK_MTORC1_SIGNALING',\n",
       "       'HALLMARK_E2F_TARGETS', 'HALLMARK_MYC_TARGETS_V1',\n",
       "       'HALLMARK_MYC_TARGETS_V2',\n",
       "       'HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION',\n",
       "       'HALLMARK_INFLAMMATORY_RESPONSE', 'HALLMARK_XENOBIOTIC_METABOLISM',\n",
       "       'HALLMARK_FATTY_ACID_METABOLISM',\n",
       "       'HALLMARK_OXIDATIVE_PHOSPHORYLATION', 'HALLMARK_GLYCOLYSIS',\n",
       "       'HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY', 'HALLMARK_P53_PATHWAY',\n",
       "       'HALLMARK_UV_RESPONSE_UP', 'HALLMARK_UV_RESPONSE_DN',\n",
       "       'HALLMARK_ANGIOGENESIS', 'HALLMARK_HEME_METABOLISM',\n",
       "       'HALLMARK_COAGULATION', 'HALLMARK_IL2_STAT5_SIGNALING',\n",
       "       'HALLMARK_BILE_ACID_METABOLISM', 'HALLMARK_PEROXISOME',\n",
       "       'HALLMARK_ALLOGRAFT_REJECTION', 'HALLMARK_SPERMATOGENESIS',\n",
       "       'HALLMARK_KRAS_SIGNALING_UP', 'HALLMARK_KRAS_SIGNALING_DN',\n",
       "       'HALLMARK_PANCREAS_BETA_CELLS', 'StromalSignature',\n",
       "       'ImmuneSignature', 'b_cells', 'cd4', 'cd8', 'cytotoxic',\n",
       "       'dendritics', 'endothelial', 'eosinophils', 'fibroblasts', 'm1',\n",
       "       'm2', 'macrophages', 'mast', 'monocytes', 'neutrophils', 'nk',\n",
       "       'plasma_cells', 'tgd', 'tregs'], dtype=object)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_sets = pd.read_csv(\n",
    "    home + \"Data/3/Hallmarks_ConsensusTME_GeneSets.txt\",\n",
    "    sep = \"\\t\"\n",
    ")\n",
    "\n",
    "gene_sets.columns.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      DCN\n",
       "1    PAPPA\n",
       "2    SFRP4\n",
       "3    THBS2\n",
       "4     LY86\n",
       "Name: StromalSignature, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stroma_geneset = gene_sets[\"StromalSignature\"]\n",
    "\n",
    "stroma_geneset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.series.Series"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(stroma_geneset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load TCGAov gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-04-1348-01</th>\n",
       "      <th>TCGA-04-1357-01</th>\n",
       "      <th>TCGA-04-1362-01</th>\n",
       "      <th>TCGA-04-1364-01</th>\n",
       "      <th>TCGA-04-1365-01</th>\n",
       "      <th>TCGA-04-1514-01</th>\n",
       "      <th>TCGA-04-1519-01</th>\n",
       "      <th>TCGA-09-0364-01</th>\n",
       "      <th>TCGA-09-0366-01</th>\n",
       "      <th>TCGA-09-0367-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-61-2102-01</th>\n",
       "      <th>TCGA-61-2104-01</th>\n",
       "      <th>TCGA-61-2109-01</th>\n",
       "      <th>TCGA-61-2110-01</th>\n",
       "      <th>TCGA-61-2111-01</th>\n",
       "      <th>TCGA-61-2113-01</th>\n",
       "      <th>TCGA-OY-A56P-01</th>\n",
       "      <th>TCGA-OY-A56Q-01</th>\n",
       "      <th>TCGA-VG-A8LO-01</th>\n",
       "      <th>TCGA-WR-A838-01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>66.4695</td>\n",
       "      <td>65.5664</td>\n",
       "      <td>41.6412</td>\n",
       "      <td>187.0368</td>\n",
       "      <td>23.9295</td>\n",
       "      <td>32.8123</td>\n",
       "      <td>46.3418</td>\n",
       "      <td>162.8522</td>\n",
       "      <td>88.5518</td>\n",
       "      <td>192.7354</td>\n",
       "      <td>...</td>\n",
       "      <td>63.7012</td>\n",
       "      <td>99.1485</td>\n",
       "      <td>38.3684</td>\n",
       "      <td>140.3668</td>\n",
       "      <td>94.7103</td>\n",
       "      <td>70.4394</td>\n",
       "      <td>43.1736</td>\n",
       "      <td>33.6854</td>\n",
       "      <td>27.9604</td>\n",
       "      <td>16.2092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>36.3243</td>\n",
       "      <td>28.1315</td>\n",
       "      <td>23.2465</td>\n",
       "      <td>114.6008</td>\n",
       "      <td>10.4957</td>\n",
       "      <td>20.1900</td>\n",
       "      <td>17.2007</td>\n",
       "      <td>83.0906</td>\n",
       "      <td>65.6758</td>\n",
       "      <td>55.1265</td>\n",
       "      <td>...</td>\n",
       "      <td>30.3602</td>\n",
       "      <td>71.1797</td>\n",
       "      <td>26.0239</td>\n",
       "      <td>77.2415</td>\n",
       "      <td>61.4890</td>\n",
       "      <td>106.1623</td>\n",
       "      <td>47.8864</td>\n",
       "      <td>25.3482</td>\n",
       "      <td>17.9853</td>\n",
       "      <td>42.9450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3310</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9034</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.9057</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>5899.8279</td>\n",
       "      <td>9384.4401</td>\n",
       "      <td>3350.4207</td>\n",
       "      <td>1455.2316</td>\n",
       "      <td>3999.3792</td>\n",
       "      <td>3224.5797</td>\n",
       "      <td>1986.1593</td>\n",
       "      <td>1487.3247</td>\n",
       "      <td>10557.8325</td>\n",
       "      <td>4603.9013</td>\n",
       "      <td>...</td>\n",
       "      <td>10264.1373</td>\n",
       "      <td>7220.7664</td>\n",
       "      <td>7200.1144</td>\n",
       "      <td>1819.9428</td>\n",
       "      <td>6179.7358</td>\n",
       "      <td>5365.7972</td>\n",
       "      <td>8773.5936</td>\n",
       "      <td>10006.7877</td>\n",
       "      <td>1888.1746</td>\n",
       "      <td>4205.6055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>118.4566</td>\n",
       "      <td>111.0026</td>\n",
       "      <td>71.3613</td>\n",
       "      <td>67.8607</td>\n",
       "      <td>52.1501</td>\n",
       "      <td>224.7582</td>\n",
       "      <td>182.0650</td>\n",
       "      <td>120.2838</td>\n",
       "      <td>53.6419</td>\n",
       "      <td>92.9112</td>\n",
       "      <td>...</td>\n",
       "      <td>146.6130</td>\n",
       "      <td>45.8648</td>\n",
       "      <td>64.4638</td>\n",
       "      <td>28.5565</td>\n",
       "      <td>100.6605</td>\n",
       "      <td>23.5625</td>\n",
       "      <td>84.2153</td>\n",
       "      <td>582.3492</td>\n",
       "      <td>59.8084</td>\n",
       "      <td>143.4910</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 307 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            TCGA-04-1348-01  TCGA-04-1357-01  TCGA-04-1362-01  \\\n",
       "GeneSymbol                                                      \n",
       "A1BG                66.4695          65.5664          41.6412   \n",
       "A1BG-AS1            36.3243          28.1315          23.2465   \n",
       "A1CF                 0.0000           0.0000           0.3310   \n",
       "A2M               5899.8279        9384.4401        3350.4207   \n",
       "A2M-AS1            118.4566         111.0026          71.3613   \n",
       "\n",
       "            TCGA-04-1364-01  TCGA-04-1365-01  TCGA-04-1514-01  \\\n",
       "GeneSymbol                                                      \n",
       "A1BG               187.0368          23.9295          32.8123   \n",
       "A1BG-AS1           114.6008          10.4957          20.1900   \n",
       "A1CF                 0.0000           0.0000           0.0000   \n",
       "A2M               1455.2316        3999.3792        3224.5797   \n",
       "A2M-AS1             67.8607          52.1501         224.7582   \n",
       "\n",
       "            TCGA-04-1519-01  TCGA-09-0364-01  TCGA-09-0366-01  \\\n",
       "GeneSymbol                                                      \n",
       "A1BG                46.3418         162.8522          88.5518   \n",
       "A1BG-AS1            17.2007          83.0906          65.6758   \n",
       "A1CF                 0.0000           0.9034           0.0000   \n",
       "A2M               1986.1593        1487.3247       10557.8325   \n",
       "A2M-AS1            182.0650         120.2838          53.6419   \n",
       "\n",
       "            TCGA-09-0367-01  ...  TCGA-61-2102-01  TCGA-61-2104-01  \\\n",
       "GeneSymbol                   ...                                     \n",
       "A1BG               192.7354  ...          63.7012          99.1485   \n",
       "A1BG-AS1            55.1265  ...          30.3602          71.1797   \n",
       "A1CF                 0.0000  ...           0.0000           0.0000   \n",
       "A2M               4603.9013  ...       10264.1373        7220.7664   \n",
       "A2M-AS1             92.9112  ...         146.6130          45.8648   \n",
       "\n",
       "            TCGA-61-2109-01  TCGA-61-2110-01  TCGA-61-2111-01  \\\n",
       "GeneSymbol                                                      \n",
       "A1BG                38.3684         140.3668          94.7103   \n",
       "A1BG-AS1            26.0239          77.2415          61.4890   \n",
       "A1CF                 0.0000           1.9057           0.0000   \n",
       "A2M               7200.1144        1819.9428        6179.7358   \n",
       "A2M-AS1             64.4638          28.5565         100.6605   \n",
       "\n",
       "            TCGA-61-2113-01  TCGA-OY-A56P-01  TCGA-OY-A56Q-01  \\\n",
       "GeneSymbol                                                      \n",
       "A1BG                70.4394          43.1736          33.6854   \n",
       "A1BG-AS1           106.1623          47.8864          25.3482   \n",
       "A1CF                 0.0000           0.0000           0.0000   \n",
       "A2M               5365.7972        8773.5936       10006.7877   \n",
       "A2M-AS1             23.5625          84.2153         582.3492   \n",
       "\n",
       "            TCGA-VG-A8LO-01  TCGA-WR-A838-01  \n",
       "GeneSymbol                                    \n",
       "A1BG                27.9604          16.2092  \n",
       "A1BG-AS1            17.9853          42.9450  \n",
       "A1CF                 0.0000           0.0000  \n",
       "A2M               1888.1746        4205.6055  \n",
       "A2M-AS1             59.8084         143.4910  \n",
       "\n",
       "[5 rows x 307 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGAov_rna = pd.read_csv(\n",
    "    home + \"Data/1/TCGAov_RNAseq.txt\",\n",
    "    sep = \"\\t\",\n",
    "    index_col = \"GeneSymbol\"\n",
    ")\n",
    "\n",
    "TCGAov_rna.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load TCGAov ABSOLUTE tumour cellularity values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>call status</th>\n",
       "      <th>purity</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>Genome doublings</th>\n",
       "      <th>Coverage for 80% power</th>\n",
       "      <th>Cancer DNA fraction</th>\n",
       "      <th>Subclonal genome fraction</th>\n",
       "      <th>solution</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>array</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J1-01</th>\n",
       "      <td>TCGA-OR-A5J1-01A-11D-A29H-01</td>\n",
       "      <td>called</td>\n",
       "      <td>0.90</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.02</td>\n",
       "      <td>new</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J2-01</th>\n",
       "      <td>TCGA-OR-A5J2-01A-11D-A29H-01</td>\n",
       "      <td>called</td>\n",
       "      <td>0.89</td>\n",
       "      <td>1.30</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.16</td>\n",
       "      <td>new</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J3-01</th>\n",
       "      <td>TCGA-OR-A5J3-01A-11D-A29H-01</td>\n",
       "      <td>called</td>\n",
       "      <td>0.93</td>\n",
       "      <td>1.27</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.11</td>\n",
       "      <td>new</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J4-01</th>\n",
       "      <td>TCGA-OR-A5J4-01A-11D-A29H-01</td>\n",
       "      <td>called</td>\n",
       "      <td>0.87</td>\n",
       "      <td>2.60</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.08</td>\n",
       "      <td>new</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J5-01</th>\n",
       "      <td>TCGA-OR-A5J5-01A-11D-A29H-01</td>\n",
       "      <td>called</td>\n",
       "      <td>0.93</td>\n",
       "      <td>2.79</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.15</td>\n",
       "      <td>new</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       sample call status  purity  ploidy  \\\n",
       "array                                                                       \n",
       "TCGA-OR-A5J1-01  TCGA-OR-A5J1-01A-11D-A29H-01      called    0.90    2.00   \n",
       "TCGA-OR-A5J2-01  TCGA-OR-A5J2-01A-11D-A29H-01      called    0.89    1.30   \n",
       "TCGA-OR-A5J3-01  TCGA-OR-A5J3-01A-11D-A29H-01      called    0.93    1.27   \n",
       "TCGA-OR-A5J4-01  TCGA-OR-A5J4-01A-11D-A29H-01      called    0.87    2.60   \n",
       "TCGA-OR-A5J5-01  TCGA-OR-A5J5-01A-11D-A29H-01      called    0.93    2.79   \n",
       "\n",
       "                 Genome doublings  Coverage for 80% power  \\\n",
       "array                                                       \n",
       "TCGA-OR-A5J1-01               0.0                     9.0   \n",
       "TCGA-OR-A5J2-01               0.0                     6.0   \n",
       "TCGA-OR-A5J3-01               0.0                     5.0   \n",
       "TCGA-OR-A5J4-01               1.0                    12.0   \n",
       "TCGA-OR-A5J5-01               1.0                    12.0   \n",
       "\n",
       "                 Cancer DNA fraction  Subclonal genome fraction solution  \n",
       "array                                                                     \n",
       "TCGA-OR-A5J1-01                 0.90                       0.02      new  \n",
       "TCGA-OR-A5J2-01                 0.84                       0.16      new  \n",
       "TCGA-OR-A5J3-01                 0.89                       0.11      new  \n",
       "TCGA-OR-A5J4-01                 0.89                       0.08      new  \n",
       "TCGA-OR-A5J5-01                 0.95                       0.15      new  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_ABSOLUTE = home + 'Data/3/TCGA_mastercalls.abs_tables_JSedit.fixed.txt'\n",
    "\n",
    "TCGA_ABSOLUTE = pd.read_csv(TCGA_ABSOLUTE,\n",
    "                            sep='\\t',\n",
    "                            index_col='array')\n",
    "\n",
    "TCGA_ABSOLUTE.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select purity (tumour cellularity) column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array\n",
       "TCGA-OR-A5J1-01    0.90\n",
       "TCGA-OR-A5J2-01    0.89\n",
       "TCGA-OR-A5J3-01    0.93\n",
       "TCGA-OR-A5J4-01    0.87\n",
       "TCGA-OR-A5J5-01    0.93\n",
       "Name: purity, dtype: float64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_ABSOLUTE = TCGA_ABSOLUTE.purity\n",
    "\n",
    "TCGA_ABSOLUTE.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Correlate gene expression with tumour cellularity\n",
    "\n",
    "### For each gene calculate its  Pearson's correlation with tumour cellularity across samples\n",
    "\n",
    "#### NOTE: Pearson correlation as TIMER has done for immune related genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1348-01</th>\n",
       "      <td>66.4695</td>\n",
       "      <td>36.3243</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5899.8279</td>\n",
       "      <td>118.4566</td>\n",
       "      <td>7.5289</td>\n",
       "      <td>92.4980</td>\n",
       "      <td>0.5378</td>\n",
       "      <td>2186.0715</td>\n",
       "      <td>1151.1159</td>\n",
       "      <td>...</td>\n",
       "      <td>928.9002</td>\n",
       "      <td>794.5684</td>\n",
       "      <td>36.0312</td>\n",
       "      <td>235.2783</td>\n",
       "      <td>827.1041</td>\n",
       "      <td>5.6467</td>\n",
       "      <td>560.0968</td>\n",
       "      <td>15871.2019</td>\n",
       "      <td>505.7811</td>\n",
       "      <td>475.9344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1357-01</th>\n",
       "      <td>65.5664</td>\n",
       "      <td>28.1315</td>\n",
       "      <td>0.000</td>\n",
       "      <td>9384.4401</td>\n",
       "      <td>111.0026</td>\n",
       "      <td>54.6875</td>\n",
       "      <td>298.1771</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1073.5677</td>\n",
       "      <td>249.3490</td>\n",
       "      <td>...</td>\n",
       "      <td>195.1107</td>\n",
       "      <td>217.4479</td>\n",
       "      <td>110.0260</td>\n",
       "      <td>583.3333</td>\n",
       "      <td>1242.1875</td>\n",
       "      <td>3.9062</td>\n",
       "      <td>600.2604</td>\n",
       "      <td>5378.9062</td>\n",
       "      <td>805.3385</td>\n",
       "      <td>415.3646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1362-01</th>\n",
       "      <td>41.6412</td>\n",
       "      <td>23.2465</td>\n",
       "      <td>0.331</td>\n",
       "      <td>3350.4207</td>\n",
       "      <td>71.3613</td>\n",
       "      <td>5.6263</td>\n",
       "      <td>697.9919</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1140.8147</td>\n",
       "      <td>581.1635</td>\n",
       "      <td>...</td>\n",
       "      <td>421.6182</td>\n",
       "      <td>605.3235</td>\n",
       "      <td>93.3304</td>\n",
       "      <td>788.6746</td>\n",
       "      <td>1575.0325</td>\n",
       "      <td>56.5939</td>\n",
       "      <td>915.7627</td>\n",
       "      <td>6137.2982</td>\n",
       "      <td>901.5315</td>\n",
       "      <td>803.8987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1364-01</th>\n",
       "      <td>187.0368</td>\n",
       "      <td>114.6008</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1455.2316</td>\n",
       "      <td>67.8607</td>\n",
       "      <td>5.0883</td>\n",
       "      <td>147.9842</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1607.4730</td>\n",
       "      <td>785.7155</td>\n",
       "      <td>...</td>\n",
       "      <td>913.9953</td>\n",
       "      <td>1079.9878</td>\n",
       "      <td>13.9928</td>\n",
       "      <td>137.3836</td>\n",
       "      <td>1138.5030</td>\n",
       "      <td>13.1447</td>\n",
       "      <td>811.5809</td>\n",
       "      <td>5972.3706</td>\n",
       "      <td>372.7166</td>\n",
       "      <td>444.8006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1365-01</th>\n",
       "      <td>23.9295</td>\n",
       "      <td>10.4957</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3999.3792</td>\n",
       "      <td>52.1501</td>\n",
       "      <td>3.3148</td>\n",
       "      <td>203.1038</td>\n",
       "      <td>0.6027</td>\n",
       "      <td>535.7842</td>\n",
       "      <td>928.7329</td>\n",
       "      <td>...</td>\n",
       "      <td>646.5632</td>\n",
       "      <td>1882.7784</td>\n",
       "      <td>35.8596</td>\n",
       "      <td>328.4617</td>\n",
       "      <td>1455.7782</td>\n",
       "      <td>7.2322</td>\n",
       "      <td>1031.7915</td>\n",
       "      <td>7211.9934</td>\n",
       "      <td>514.0877</td>\n",
       "      <td>787.1026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 20437 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GeneSymbol           A1BG  A1BG-AS1   A1CF        A2M   A2M-AS1    A2ML1  \\\n",
       "TCGA-04-1348-01   66.4695   36.3243  0.000  5899.8279  118.4566   7.5289   \n",
       "TCGA-04-1357-01   65.5664   28.1315  0.000  9384.4401  111.0026  54.6875   \n",
       "TCGA-04-1362-01   41.6412   23.2465  0.331  3350.4207   71.3613   5.6263   \n",
       "TCGA-04-1364-01  187.0368  114.6008  0.000  1455.2316   67.8607   5.0883   \n",
       "TCGA-04-1365-01   23.9295   10.4957  0.000  3999.3792   52.1501   3.3148   \n",
       "\n",
       "GeneSymbol         A4GALT   A4GNT       AAAS       AACS  ...    ZWILCH  \\\n",
       "TCGA-04-1348-01   92.4980  0.5378  2186.0715  1151.1159  ...  928.9002   \n",
       "TCGA-04-1357-01  298.1771  0.0000  1073.5677   249.3490  ...  195.1107   \n",
       "TCGA-04-1362-01  697.9919  0.0000  1140.8147   581.1635  ...  421.6182   \n",
       "TCGA-04-1364-01  147.9842  0.0000  1607.4730   785.7155  ...  913.9953   \n",
       "TCGA-04-1365-01  203.1038  0.6027   535.7842   928.7329  ...  646.5632   \n",
       "\n",
       "GeneSymbol           ZWINT      ZXDA      ZXDB       ZXDC   ZYG11A     ZYG11B  \\\n",
       "TCGA-04-1348-01   794.5684   36.0312  235.2783   827.1041   5.6467   560.0968   \n",
       "TCGA-04-1357-01   217.4479  110.0260  583.3333  1242.1875   3.9062   600.2604   \n",
       "TCGA-04-1362-01   605.3235   93.3304  788.6746  1575.0325  56.5939   915.7627   \n",
       "TCGA-04-1364-01  1079.9878   13.9928  137.3836  1138.5030  13.1447   811.5809   \n",
       "TCGA-04-1365-01  1882.7784   35.8596  328.4617  1455.7782   7.2322  1031.7915   \n",
       "\n",
       "GeneSymbol              ZYX     ZZEF1      ZZZ3  \n",
       "TCGA-04-1348-01  15871.2019  505.7811  475.9344  \n",
       "TCGA-04-1357-01   5378.9062  805.3385  415.3646  \n",
       "TCGA-04-1362-01   6137.2982  901.5315  803.8987  \n",
       "TCGA-04-1364-01   5972.3706  372.7166  444.8006  \n",
       "TCGA-04-1365-01   7211.9934  514.0877  787.1026  \n",
       "\n",
       "[5 rows x 20437 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGAov_rna = TCGAov_rna.T\n",
    "\n",
    "TCGAov_rna.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select only stromal genes in the TCGAov gene expression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>C1QB</th>\n",
       "      <th>C1QA</th>\n",
       "      <th>COL8A2</th>\n",
       "      <th>ENPEP</th>\n",
       "      <th>GIMAP5</th>\n",
       "      <th>FRZB</th>\n",
       "      <th>SERPING1</th>\n",
       "      <th>COL10A1</th>\n",
       "      <th>DCN</th>\n",
       "      <th>ISLR</th>\n",
       "      <th>...</th>\n",
       "      <th>VCAM1</th>\n",
       "      <th>ARHGAP28</th>\n",
       "      <th>AOC3</th>\n",
       "      <th>CD33</th>\n",
       "      <th>ZNF423</th>\n",
       "      <th>WISP1</th>\n",
       "      <th>RGS4</th>\n",
       "      <th>MAF</th>\n",
       "      <th>ECM2</th>\n",
       "      <th>TLR7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1348-01</th>\n",
       "      <td>10164.2915</td>\n",
       "      <td>6170.2070</td>\n",
       "      <td>165.9048</td>\n",
       "      <td>177.7360</td>\n",
       "      <td>418.9298</td>\n",
       "      <td>114.2780</td>\n",
       "      <td>19434.2565</td>\n",
       "      <td>118.5803</td>\n",
       "      <td>1644.7970</td>\n",
       "      <td>1044.9045</td>\n",
       "      <td>...</td>\n",
       "      <td>671.1482</td>\n",
       "      <td>38.1823</td>\n",
       "      <td>27.9645</td>\n",
       "      <td>108.6314</td>\n",
       "      <td>114.2780</td>\n",
       "      <td>54.8535</td>\n",
       "      <td>86.0446</td>\n",
       "      <td>542.3501</td>\n",
       "      <td>55.6601</td>\n",
       "      <td>116.9669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1357-01</th>\n",
       "      <td>19589.1927</td>\n",
       "      <td>10033.2031</td>\n",
       "      <td>209.6354</td>\n",
       "      <td>88.5417</td>\n",
       "      <td>671.8750</td>\n",
       "      <td>154.2969</td>\n",
       "      <td>19902.3438</td>\n",
       "      <td>825.5208</td>\n",
       "      <td>4611.3281</td>\n",
       "      <td>2503.2552</td>\n",
       "      <td>...</td>\n",
       "      <td>521.4844</td>\n",
       "      <td>29.2969</td>\n",
       "      <td>77.4740</td>\n",
       "      <td>149.0885</td>\n",
       "      <td>318.3594</td>\n",
       "      <td>68.3594</td>\n",
       "      <td>109.3750</td>\n",
       "      <td>1046.2240</td>\n",
       "      <td>91.1458</td>\n",
       "      <td>421.2240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1362-01</th>\n",
       "      <td>3141.1291</td>\n",
       "      <td>1940.0799</td>\n",
       "      <td>174.4152</td>\n",
       "      <td>93.3304</td>\n",
       "      <td>96.3090</td>\n",
       "      <td>56.9249</td>\n",
       "      <td>6737.9883</td>\n",
       "      <td>18.5337</td>\n",
       "      <td>1208.9921</td>\n",
       "      <td>452.0896</td>\n",
       "      <td>...</td>\n",
       "      <td>56.2630</td>\n",
       "      <td>26.1457</td>\n",
       "      <td>35.0816</td>\n",
       "      <td>40.0460</td>\n",
       "      <td>223.3971</td>\n",
       "      <td>30.7792</td>\n",
       "      <td>55.2701</td>\n",
       "      <td>338.2398</td>\n",
       "      <td>32.1030</td>\n",
       "      <td>150.2553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1364-01</th>\n",
       "      <td>144.5920</td>\n",
       "      <td>137.8076</td>\n",
       "      <td>44.5225</td>\n",
       "      <td>79.7164</td>\n",
       "      <td>30.9537</td>\n",
       "      <td>159.0088</td>\n",
       "      <td>9896.7079</td>\n",
       "      <td>4.2402</td>\n",
       "      <td>996.8792</td>\n",
       "      <td>296.3924</td>\n",
       "      <td>...</td>\n",
       "      <td>49.1867</td>\n",
       "      <td>183.1781</td>\n",
       "      <td>424.0235</td>\n",
       "      <td>2.1201</td>\n",
       "      <td>585.1524</td>\n",
       "      <td>14.4168</td>\n",
       "      <td>49.6107</td>\n",
       "      <td>100.4936</td>\n",
       "      <td>39.4342</td>\n",
       "      <td>7.2084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1365-01</th>\n",
       "      <td>4292.9034</td>\n",
       "      <td>3111.9482</td>\n",
       "      <td>108.4827</td>\n",
       "      <td>179.5992</td>\n",
       "      <td>276.0283</td>\n",
       "      <td>1077.2940</td>\n",
       "      <td>11157.7520</td>\n",
       "      <td>34.3529</td>\n",
       "      <td>1119.4817</td>\n",
       "      <td>390.2366</td>\n",
       "      <td>...</td>\n",
       "      <td>178.9965</td>\n",
       "      <td>73.5272</td>\n",
       "      <td>45.8038</td>\n",
       "      <td>65.9937</td>\n",
       "      <td>68.1031</td>\n",
       "      <td>44.8998</td>\n",
       "      <td>91.3063</td>\n",
       "      <td>276.0283</td>\n",
       "      <td>26.2167</td>\n",
       "      <td>71.1165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 137 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GeneSymbol             C1QB        C1QA    COL8A2     ENPEP    GIMAP5  \\\n",
       "TCGA-04-1348-01  10164.2915   6170.2070  165.9048  177.7360  418.9298   \n",
       "TCGA-04-1357-01  19589.1927  10033.2031  209.6354   88.5417  671.8750   \n",
       "TCGA-04-1362-01   3141.1291   1940.0799  174.4152   93.3304   96.3090   \n",
       "TCGA-04-1364-01    144.5920    137.8076   44.5225   79.7164   30.9537   \n",
       "TCGA-04-1365-01   4292.9034   3111.9482  108.4827  179.5992  276.0283   \n",
       "\n",
       "GeneSymbol            FRZB    SERPING1   COL10A1        DCN       ISLR  ...  \\\n",
       "TCGA-04-1348-01   114.2780  19434.2565  118.5803  1644.7970  1044.9045  ...   \n",
       "TCGA-04-1357-01   154.2969  19902.3438  825.5208  4611.3281  2503.2552  ...   \n",
       "TCGA-04-1362-01    56.9249   6737.9883   18.5337  1208.9921   452.0896  ...   \n",
       "TCGA-04-1364-01   159.0088   9896.7079    4.2402   996.8792   296.3924  ...   \n",
       "TCGA-04-1365-01  1077.2940  11157.7520   34.3529  1119.4817   390.2366  ...   \n",
       "\n",
       "GeneSymbol          VCAM1  ARHGAP28      AOC3      CD33    ZNF423    WISP1  \\\n",
       "TCGA-04-1348-01  671.1482   38.1823   27.9645  108.6314  114.2780  54.8535   \n",
       "TCGA-04-1357-01  521.4844   29.2969   77.4740  149.0885  318.3594  68.3594   \n",
       "TCGA-04-1362-01   56.2630   26.1457   35.0816   40.0460  223.3971  30.7792   \n",
       "TCGA-04-1364-01   49.1867  183.1781  424.0235    2.1201  585.1524  14.4168   \n",
       "TCGA-04-1365-01  178.9965   73.5272   45.8038   65.9937   68.1031  44.8998   \n",
       "\n",
       "GeneSymbol           RGS4        MAF     ECM2      TLR7  \n",
       "TCGA-04-1348-01   86.0446   542.3501  55.6601  116.9669  \n",
       "TCGA-04-1357-01  109.3750  1046.2240  91.1458  421.2240  \n",
       "TCGA-04-1362-01   55.2701   338.2398  32.1030  150.2553  \n",
       "TCGA-04-1364-01   49.6107   100.4936  39.4342    7.2084  \n",
       "TCGA-04-1365-01   91.3063   276.0283  26.2167   71.1165  \n",
       "\n",
       "[5 rows x 137 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGAov_rna_stromal = TCGAov_rna[set(TCGAov_rna.columns.values) & set(stroma_geneset)]\n",
    "\n",
    "TCGAov_rna_stromal.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select only TCGAov samples for ABSOLUTE tumour cellularity scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array\n",
       "TCGA-04-1348-01    0.76\n",
       "TCGA-04-1357-01    0.52\n",
       "TCGA-04-1362-01    0.86\n",
       "TCGA-09-0364-01    0.89\n",
       "TCGA-09-0369-01    0.90\n",
       "Name: purity, dtype: float64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGAov_ABSOLUTE = TCGA_ABSOLUTE[list(TCGA_ABSOLUTE.index & TCGAov_rna_stromal.index)]\n",
    "\n",
    "TCGAov_ABSOLUTE.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGAov_ABSOLUTE.isnull().values.any()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove samples that do not have tumour cellularity data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "TCGAov_ABSOLUTE = TCGAov_ABSOLUTE[~TCGAov_ABSOLUTE.isnull()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Match expression matrix samples to tumour cellularity data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>C1QB</th>\n",
       "      <th>C1QA</th>\n",
       "      <th>COL8A2</th>\n",
       "      <th>ENPEP</th>\n",
       "      <th>GIMAP5</th>\n",
       "      <th>FRZB</th>\n",
       "      <th>SERPING1</th>\n",
       "      <th>COL10A1</th>\n",
       "      <th>DCN</th>\n",
       "      <th>ISLR</th>\n",
       "      <th>...</th>\n",
       "      <th>VCAM1</th>\n",
       "      <th>ARHGAP28</th>\n",
       "      <th>AOC3</th>\n",
       "      <th>CD33</th>\n",
       "      <th>ZNF423</th>\n",
       "      <th>WISP1</th>\n",
       "      <th>RGS4</th>\n",
       "      <th>MAF</th>\n",
       "      <th>ECM2</th>\n",
       "      <th>TLR7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>array</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1348-01</th>\n",
       "      <td>10164.2915</td>\n",
       "      <td>6170.2070</td>\n",
       "      <td>165.9048</td>\n",
       "      <td>177.7360</td>\n",
       "      <td>418.9298</td>\n",
       "      <td>114.2780</td>\n",
       "      <td>19434.2565</td>\n",
       "      <td>118.5803</td>\n",
       "      <td>1644.7970</td>\n",
       "      <td>1044.9045</td>\n",
       "      <td>...</td>\n",
       "      <td>671.1482</td>\n",
       "      <td>38.1823</td>\n",
       "      <td>27.9645</td>\n",
       "      <td>108.6314</td>\n",
       "      <td>114.2780</td>\n",
       "      <td>54.8535</td>\n",
       "      <td>86.0446</td>\n",
       "      <td>542.3501</td>\n",
       "      <td>55.6601</td>\n",
       "      <td>116.9669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1357-01</th>\n",
       "      <td>19589.1927</td>\n",
       "      <td>10033.2031</td>\n",
       "      <td>209.6354</td>\n",
       "      <td>88.5417</td>\n",
       "      <td>671.8750</td>\n",
       "      <td>154.2969</td>\n",
       "      <td>19902.3438</td>\n",
       "      <td>825.5208</td>\n",
       "      <td>4611.3281</td>\n",
       "      <td>2503.2552</td>\n",
       "      <td>...</td>\n",
       "      <td>521.4844</td>\n",
       "      <td>29.2969</td>\n",
       "      <td>77.4740</td>\n",
       "      <td>149.0885</td>\n",
       "      <td>318.3594</td>\n",
       "      <td>68.3594</td>\n",
       "      <td>109.3750</td>\n",
       "      <td>1046.2240</td>\n",
       "      <td>91.1458</td>\n",
       "      <td>421.2240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1362-01</th>\n",
       "      <td>3141.1291</td>\n",
       "      <td>1940.0799</td>\n",
       "      <td>174.4152</td>\n",
       "      <td>93.3304</td>\n",
       "      <td>96.3090</td>\n",
       "      <td>56.9249</td>\n",
       "      <td>6737.9883</td>\n",
       "      <td>18.5337</td>\n",
       "      <td>1208.9921</td>\n",
       "      <td>452.0896</td>\n",
       "      <td>...</td>\n",
       "      <td>56.2630</td>\n",
       "      <td>26.1457</td>\n",
       "      <td>35.0816</td>\n",
       "      <td>40.0460</td>\n",
       "      <td>223.3971</td>\n",
       "      <td>30.7792</td>\n",
       "      <td>55.2701</td>\n",
       "      <td>338.2398</td>\n",
       "      <td>32.1030</td>\n",
       "      <td>150.2553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-09-0364-01</th>\n",
       "      <td>442.2165</td>\n",
       "      <td>279.1520</td>\n",
       "      <td>39.7498</td>\n",
       "      <td>136.4141</td>\n",
       "      <td>28.0055</td>\n",
       "      <td>460.7363</td>\n",
       "      <td>9112.6421</td>\n",
       "      <td>4.0653</td>\n",
       "      <td>1506.8789</td>\n",
       "      <td>630.1247</td>\n",
       "      <td>...</td>\n",
       "      <td>402.0150</td>\n",
       "      <td>261.5356</td>\n",
       "      <td>116.9909</td>\n",
       "      <td>8.1306</td>\n",
       "      <td>7247.1116</td>\n",
       "      <td>43.8151</td>\n",
       "      <td>46.0736</td>\n",
       "      <td>1014.0717</td>\n",
       "      <td>68.2070</td>\n",
       "      <td>5.4204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-09-0369-01</th>\n",
       "      <td>3830.2561</td>\n",
       "      <td>2882.4549</td>\n",
       "      <td>305.3635</td>\n",
       "      <td>409.8917</td>\n",
       "      <td>149.1583</td>\n",
       "      <td>381.7043</td>\n",
       "      <td>12493.4762</td>\n",
       "      <td>14.0937</td>\n",
       "      <td>18318.5778</td>\n",
       "      <td>11582.0837</td>\n",
       "      <td>...</td>\n",
       "      <td>326.5040</td>\n",
       "      <td>550.8287</td>\n",
       "      <td>343.8275</td>\n",
       "      <td>64.8897</td>\n",
       "      <td>859.1283</td>\n",
       "      <td>134.4774</td>\n",
       "      <td>214.3417</td>\n",
       "      <td>1424.0507</td>\n",
       "      <td>87.2048</td>\n",
       "      <td>45.5109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 137 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GeneSymbol             C1QB        C1QA    COL8A2     ENPEP    GIMAP5  \\\n",
       "array                                                                   \n",
       "TCGA-04-1348-01  10164.2915   6170.2070  165.9048  177.7360  418.9298   \n",
       "TCGA-04-1357-01  19589.1927  10033.2031  209.6354   88.5417  671.8750   \n",
       "TCGA-04-1362-01   3141.1291   1940.0799  174.4152   93.3304   96.3090   \n",
       "TCGA-09-0364-01    442.2165    279.1520   39.7498  136.4141   28.0055   \n",
       "TCGA-09-0369-01   3830.2561   2882.4549  305.3635  409.8917  149.1583   \n",
       "\n",
       "GeneSymbol           FRZB    SERPING1   COL10A1         DCN        ISLR  ...  \\\n",
       "array                                                                    ...   \n",
       "TCGA-04-1348-01  114.2780  19434.2565  118.5803   1644.7970   1044.9045  ...   \n",
       "TCGA-04-1357-01  154.2969  19902.3438  825.5208   4611.3281   2503.2552  ...   \n",
       "TCGA-04-1362-01   56.9249   6737.9883   18.5337   1208.9921    452.0896  ...   \n",
       "TCGA-09-0364-01  460.7363   9112.6421    4.0653   1506.8789    630.1247  ...   \n",
       "TCGA-09-0369-01  381.7043  12493.4762   14.0937  18318.5778  11582.0837  ...   \n",
       "\n",
       "GeneSymbol          VCAM1  ARHGAP28      AOC3      CD33     ZNF423     WISP1  \\\n",
       "array                                                                          \n",
       "TCGA-04-1348-01  671.1482   38.1823   27.9645  108.6314   114.2780   54.8535   \n",
       "TCGA-04-1357-01  521.4844   29.2969   77.4740  149.0885   318.3594   68.3594   \n",
       "TCGA-04-1362-01   56.2630   26.1457   35.0816   40.0460   223.3971   30.7792   \n",
       "TCGA-09-0364-01  402.0150  261.5356  116.9909    8.1306  7247.1116   43.8151   \n",
       "TCGA-09-0369-01  326.5040  550.8287  343.8275   64.8897   859.1283  134.4774   \n",
       "\n",
       "GeneSymbol           RGS4        MAF     ECM2      TLR7  \n",
       "array                                                    \n",
       "TCGA-04-1348-01   86.0446   542.3501  55.6601  116.9669  \n",
       "TCGA-04-1357-01  109.3750  1046.2240  91.1458  421.2240  \n",
       "TCGA-04-1362-01   55.2701   338.2398  32.1030  150.2553  \n",
       "TCGA-09-0364-01   46.0736  1014.0717  68.2070    5.4204  \n",
       "TCGA-09-0369-01  214.3417  1424.0507  87.2048   45.5109  \n",
       "\n",
       "[5 rows x 137 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGAov_rna_stromal = TCGAov_rna_stromal.loc[TCGAov_ABSOLUTE.index]\n",
    "\n",
    "TCGAov_rna_stromal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(293, 137)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGAov_rna_stromal.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>C1QB</th>\n",
       "      <th>C1QA</th>\n",
       "      <th>COL8A2</th>\n",
       "      <th>ENPEP</th>\n",
       "      <th>GIMAP5</th>\n",
       "      <th>FRZB</th>\n",
       "      <th>SERPING1</th>\n",
       "      <th>COL10A1</th>\n",
       "      <th>DCN</th>\n",
       "      <th>ISLR</th>\n",
       "      <th>...</th>\n",
       "      <th>VCAM1</th>\n",
       "      <th>ARHGAP28</th>\n",
       "      <th>AOC3</th>\n",
       "      <th>CD33</th>\n",
       "      <th>ZNF423</th>\n",
       "      <th>WISP1</th>\n",
       "      <th>RGS4</th>\n",
       "      <th>MAF</th>\n",
       "      <th>ECM2</th>\n",
       "      <th>TLR7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>array</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1348-01</th>\n",
       "      <td>10164.2915</td>\n",
       "      <td>6170.2070</td>\n",
       "      <td>165.9048</td>\n",
       "      <td>177.7360</td>\n",
       "      <td>418.9298</td>\n",
       "      <td>114.2780</td>\n",
       "      <td>19434.2565</td>\n",
       "      <td>118.5803</td>\n",
       "      <td>1644.7970</td>\n",
       "      <td>1044.9045</td>\n",
       "      <td>...</td>\n",
       "      <td>671.1482</td>\n",
       "      <td>38.1823</td>\n",
       "      <td>27.9645</td>\n",
       "      <td>108.6314</td>\n",
       "      <td>114.2780</td>\n",
       "      <td>54.8535</td>\n",
       "      <td>86.0446</td>\n",
       "      <td>542.3501</td>\n",
       "      <td>55.6601</td>\n",
       "      <td>116.9669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1357-01</th>\n",
       "      <td>19589.1927</td>\n",
       "      <td>10033.2031</td>\n",
       "      <td>209.6354</td>\n",
       "      <td>88.5417</td>\n",
       "      <td>671.8750</td>\n",
       "      <td>154.2969</td>\n",
       "      <td>19902.3438</td>\n",
       "      <td>825.5208</td>\n",
       "      <td>4611.3281</td>\n",
       "      <td>2503.2552</td>\n",
       "      <td>...</td>\n",
       "      <td>521.4844</td>\n",
       "      <td>29.2969</td>\n",
       "      <td>77.4740</td>\n",
       "      <td>149.0885</td>\n",
       "      <td>318.3594</td>\n",
       "      <td>68.3594</td>\n",
       "      <td>109.3750</td>\n",
       "      <td>1046.2240</td>\n",
       "      <td>91.1458</td>\n",
       "      <td>421.2240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1362-01</th>\n",
       "      <td>3141.1291</td>\n",
       "      <td>1940.0799</td>\n",
       "      <td>174.4152</td>\n",
       "      <td>93.3304</td>\n",
       "      <td>96.3090</td>\n",
       "      <td>56.9249</td>\n",
       "      <td>6737.9883</td>\n",
       "      <td>18.5337</td>\n",
       "      <td>1208.9921</td>\n",
       "      <td>452.0896</td>\n",
       "      <td>...</td>\n",
       "      <td>56.2630</td>\n",
       "      <td>26.1457</td>\n",
       "      <td>35.0816</td>\n",
       "      <td>40.0460</td>\n",
       "      <td>223.3971</td>\n",
       "      <td>30.7792</td>\n",
       "      <td>55.2701</td>\n",
       "      <td>338.2398</td>\n",
       "      <td>32.1030</td>\n",
       "      <td>150.2553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-09-0364-01</th>\n",
       "      <td>442.2165</td>\n",
       "      <td>279.1520</td>\n",
       "      <td>39.7498</td>\n",
       "      <td>136.4141</td>\n",
       "      <td>28.0055</td>\n",
       "      <td>460.7363</td>\n",
       "      <td>9112.6421</td>\n",
       "      <td>4.0653</td>\n",
       "      <td>1506.8789</td>\n",
       "      <td>630.1247</td>\n",
       "      <td>...</td>\n",
       "      <td>402.0150</td>\n",
       "      <td>261.5356</td>\n",
       "      <td>116.9909</td>\n",
       "      <td>8.1306</td>\n",
       "      <td>7247.1116</td>\n",
       "      <td>43.8151</td>\n",
       "      <td>46.0736</td>\n",
       "      <td>1014.0717</td>\n",
       "      <td>68.2070</td>\n",
       "      <td>5.4204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-09-0369-01</th>\n",
       "      <td>3830.2561</td>\n",
       "      <td>2882.4549</td>\n",
       "      <td>305.3635</td>\n",
       "      <td>409.8917</td>\n",
       "      <td>149.1583</td>\n",
       "      <td>381.7043</td>\n",
       "      <td>12493.4762</td>\n",
       "      <td>14.0937</td>\n",
       "      <td>18318.5778</td>\n",
       "      <td>11582.0837</td>\n",
       "      <td>...</td>\n",
       "      <td>326.5040</td>\n",
       "      <td>550.8287</td>\n",
       "      <td>343.8275</td>\n",
       "      <td>64.8897</td>\n",
       "      <td>859.1283</td>\n",
       "      <td>134.4774</td>\n",
       "      <td>214.3417</td>\n",
       "      <td>1424.0507</td>\n",
       "      <td>87.2048</td>\n",
       "      <td>45.5109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 137 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GeneSymbol             C1QB        C1QA    COL8A2     ENPEP    GIMAP5  \\\n",
       "array                                                                   \n",
       "TCGA-04-1348-01  10164.2915   6170.2070  165.9048  177.7360  418.9298   \n",
       "TCGA-04-1357-01  19589.1927  10033.2031  209.6354   88.5417  671.8750   \n",
       "TCGA-04-1362-01   3141.1291   1940.0799  174.4152   93.3304   96.3090   \n",
       "TCGA-09-0364-01    442.2165    279.1520   39.7498  136.4141   28.0055   \n",
       "TCGA-09-0369-01   3830.2561   2882.4549  305.3635  409.8917  149.1583   \n",
       "\n",
       "GeneSymbol           FRZB    SERPING1   COL10A1         DCN        ISLR  ...  \\\n",
       "array                                                                    ...   \n",
       "TCGA-04-1348-01  114.2780  19434.2565  118.5803   1644.7970   1044.9045  ...   \n",
       "TCGA-04-1357-01  154.2969  19902.3438  825.5208   4611.3281   2503.2552  ...   \n",
       "TCGA-04-1362-01   56.9249   6737.9883   18.5337   1208.9921    452.0896  ...   \n",
       "TCGA-09-0364-01  460.7363   9112.6421    4.0653   1506.8789    630.1247  ...   \n",
       "TCGA-09-0369-01  381.7043  12493.4762   14.0937  18318.5778  11582.0837  ...   \n",
       "\n",
       "GeneSymbol          VCAM1  ARHGAP28      AOC3      CD33     ZNF423     WISP1  \\\n",
       "array                                                                          \n",
       "TCGA-04-1348-01  671.1482   38.1823   27.9645  108.6314   114.2780   54.8535   \n",
       "TCGA-04-1357-01  521.4844   29.2969   77.4740  149.0885   318.3594   68.3594   \n",
       "TCGA-04-1362-01   56.2630   26.1457   35.0816   40.0460   223.3971   30.7792   \n",
       "TCGA-09-0364-01  402.0150  261.5356  116.9909    8.1306  7247.1116   43.8151   \n",
       "TCGA-09-0369-01  326.5040  550.8287  343.8275   64.8897   859.1283  134.4774   \n",
       "\n",
       "GeneSymbol           RGS4        MAF     ECM2      TLR7  \n",
       "array                                                    \n",
       "TCGA-04-1348-01   86.0446   542.3501  55.6601  116.9669  \n",
       "TCGA-04-1357-01  109.3750  1046.2240  91.1458  421.2240  \n",
       "TCGA-04-1362-01   55.2701   338.2398  32.1030  150.2553  \n",
       "TCGA-09-0364-01   46.0736  1014.0717  68.2070    5.4204  \n",
       "TCGA-09-0369-01  214.3417  1424.0507  87.2048   45.5109  \n",
       "\n",
       "[5 rows x 137 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGAov_rna_stromal.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check patient ID correspondance between data sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "False in (TCGAov_rna_stromal.index == TCGAov_ABSOLUTE.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check nan in expression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGAov_rna_stromal.isnull().values.any()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate Spearman's correlation and keep genes with rho <= -0.2 and p-val <= 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "stromal_genes_kept = set()\n",
    "\n",
    "for stromal_gene in TCGAov_rna_stromal:\n",
    "    \n",
    "    rho, p_val = scipy.stats.spearmanr(\n",
    "        a = TCGAov_rna_stromal[stromal_gene],\n",
    "        b = TCGAov_ABSOLUTE\n",
    "    )\n",
    "    \n",
    "    if rho <= -0.2 and p_val <= 0.05:\n",
    "    \n",
    "        stromal_genes_kept.add(stromal_gene)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "125"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(stromal_genes_kept)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tumour_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C1QB</th>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1QA</th>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL8A2</th>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENPEP</th>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GIMAP5</th>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Tumour_type\n",
       "Gene_Symbol            \n",
       "C1QB                 OV\n",
       "C1QA                 OV\n",
       "COL8A2               OV\n",
       "ENPEP                OV\n",
       "GIMAP5               OV"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stromal_genes_kept_df = pd.DataFrame({\n",
    "    \"Gene_Symbol\" : pd.Series(list(stromal_genes_kept)),\n",
    "    \"Tumour_type\" : \"OV\"\n",
    "})\n",
    "\n",
    "stromal_genes_kept_df.set_index(\n",
    "    \"Gene_Symbol\",\n",
    "    inplace = True\n",
    ")\n",
    "\n",
    "stromal_genes_kept_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(125, 1)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stromal_genes_kept_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "stromal_genes_kept_df.to_csv(\n",
    "    \"../../Data/3/TCGAov_ESTIMATE_stromal_genes_keep.txt\",\n",
    "    sep = \"\\t\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# End script"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
