## DNA copy-number and mutation analysis of treatment-naive HGSOC

<img src="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/blob/master/Figures/supplementary/4/OVCT_Fig_s4_part1.png" width="100%">

<img src="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/blob/master/Figures/supplementary/4/OVCT_Fig_s4_part2.png" width="100%">

<p align="justify">
<strong>a</strong>, Absolute copy-numbers by chromosome assessed from whole-exome sequencing (WES) data using TITAN. Hierarchical clustering using euclidean distance per gene. <strong>b</strong>, Correlation between the WES-derived tumor cellularity (TITAN) and immune infiltration inferred from gene expression (ESTIMATE). Pearson correlation, two-sided t-test. <strong>c</strong>, Comparison of TITAN CNA and ASCAT tumor cellularity estimates from WES data. Kendall’s tau correlation, one-sided exact test normal approximation. <strong>d</strong>, Association between WES-derived tumor cellularity and T cell infiltration. The immune cell counts represent histology slide averages and 95% confidence intervals from immunofluorescence measurements of tumor infiltrating leukocyte (TIL) markers for CD8+, CD4+, and regulatory T cells. Kendall’s tau correlation, two-sided exact test normal approximation. <strong>e</strong>, Comparison of G2M checkpoint ssGSEA NES between samples with and without high-level (≥ 6 copies) KRAS amplification (two-sided Welch’s t-test, N = 36 samples, n = 10 independent patients). <strong>f</strong>, Correlation between G2M checkpoint ssGSEA NES and the fraction of tumor cells staining positive for the proliferation marker Ki67. Kendall’s tau correlation, two-sided exact test. <strong>g</strong>, Correlation between copy-number signature exposures derived from WES data vs SNP array using TCGA OV samples (n = 37). We first corrected exome sequencing read counts with CopywriteR, which is designed to utilize off-target reads for CNA estimation20. These corrected counts are equivalent to shallow WGS profiling and can thus be used to estimate copy number signature exposures. Note that all signature exposures correlated significantly except for signature 5, which is expected due to resolution differences between the technologies (see Supplementary Note). Kendall’s tau correlations were calculated, two-sided exact tests. <strong>h</strong>, Association between ESTIMATE ImmuneScore and the exposure of copy-number signature 4 (left) and the isometric log ratio (ILR) transform (taking into account the compositional data structure) comparing signature 4 against all other signatures (right). Kendall’s tau partial correlation with patients as a covariate, two-sided exact test. <strong>i</strong>, Similar to h except signature 1 was analyzed. Note that a positive correlation between immune infiltration and signature 1 associated with Ras signaling at first sight appears to contradict our other results, including the observations of immune exclusion associated with G2M checkpoint related gene sets (<a href="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/tree/master/Figures/main/3">Fig. 3g</a>) and G2M enrichment and Ki67 proliferation marker associated with KRAS amplifications (<a href="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/tree/master/Figures/supplementary/4">Supplementary Figs. 4e, f</a>). It is important to note that in the original copy-number signature paper, the association between Ras signalling and copy-number signature 1 was predominantly driven by KRAS G12 missense mutations and NF1 deletions, and KRAS amplifications were not associated with copy-number signature 1 (Macintyre et al. 2018, <a href="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/tree/master/Tables/supplementary">Supplementary Table 8</a>). By contrast, our cohort has a substantial proportion of samples (11 out of 50) with KRAS amplifications (≥ 6 copies), and we did not identify NF1 alterations and only one patient had a KRAS missense mutation. Thus, there are differences between the nature of KRAS alterations in the two cohorts. <strong>j</strong>, Copy-number signature 1 association with G2M checkpoint gene set enrichment. Using samples with both mRNA and WES data (n = 30), hallmark G2M NES was compared (two-sided Welch’s t-test) between samples with copy-number signature 1 exposure below (low s1) and higher than the median exposure (high s1). Samples with high-level KRAS amplification (≥ 6 copies) are indicated by red color. Note that this indicates that in our cohort we find that samples with high signature 1 exposure have significantly lower G2M gene set enrichment than samples with low exposure and this was not associated with KRAS amplification. Taken together these factors highlight the complexity of copy-number signature analysis and underscores the importance of new approaches including single cell genomic analysis to interrogate the significance of copy-number signature 1 and 4 on immune infiltration in metastatic HGSOC. <strong>k</strong>, Functional mutation enrichment and tumor cellularity associations per hallmark gene set were tested through the Chi-squared likelihood ratio test using the tumor mutation load and the patient dependency as covariates (N = 8 independent patients with n = 50 samples). Rows are sorted by BH FDR corrected P values (Benjamini, Hochberg, and Yekutieli) and columns by WES-derived tumor cellularity (TITAN). <strong>l</strong>, Functional mutation enrichment score and mRNA normalized enrichment score (NES) correlation of gene sets with mutation enrichment associated with tumor cellularity (except Apoptosis and ROS gene sets which showed no significant correlation). Kendall’s tau-b correlation coefficients and two-sided P values asymptotic tests were calculated (N = 8 independent patients with n = 36 samples). <strong>m</strong>, Gene set enrichment analysis (GSEA) of gene-level correlation between copy-numbers derived from CopywriteR and gene expression (CNA~mRNA, per-gene median n = 36 samples), compared to enrichment in correlations between gene expression and tumor cellularity (mRNA~tumor cellularity, n = 36 samples). Both Spearman’s rank correlation (see Methods). <strong>n</strong>, Left, association between copy number alterations of c-Myc (MYC) with its corresponding expression values (Log2). Spearman correlation, two-sided t-test asymptotic approximation. Box plots show median, interquartile range (25th and 75th percentiles) and 1.5x interquartile range. All samples are plotted. Middle, Notched box plots show median and interquartile range (25th and 75th percentiles), 1.5x interquartile range (top) and outliers. Middle, two- sided Welch’s t-test between low and high tumor cellularity samples (defined by median WES-derived tumor cellularity). Right, correlation between tumor cellularity and mRNA levels of c-Myc. Kendall’s tau-b correlation and two-sided P value asymptotic tests were calculated. *, †, and ‡ indicate pairs of pseudoreplicates (see <a href="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/tree/master/Figures/main/1">Fig. 1a</a>).
</p>
