## Unbiased analysis of TME heterogeneity in treatment-naive HGSOC tumors

<img align="left" src="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/blob/Figures/Figures/main/3/OVCT_Fig_m3.png" width="50%">

<p align="justify">
<strong>a</strong>, Treatment-naive cohort correlation between total TME cell estimation scores and WES-derived tumor cellularity (TITAN) (left plot) and the percentage of CD8+, CD4+ and Treg cells determined by IF staining (rightplot). <strong>b</strong>, Correlation in the TCGA ovarian cancer cohort between total TME cell estimation scores for each method and WES-derived tumor cellularity (ABSOLUTE). Fitted multiple linear regression analysis used the TCGA leukocyte methylation score as the response variable and the estimated immune cell types as explanatory variables (Supplementary Note and <a href="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/tree/Tables/Tables/supplementary">Supplementary Table 3a</a>). Adjusted R2, Akaike information criterion (AIC) and Bayesian information criterion (BIC) values were calculated to compare both goodness of fit and model simplicity. Kendall’s tau correlation coefficients and P values were calculated for a and b with the exact Kendall’s tau-b two-sided test. Normality and homoscedasticity assumptions were tested for all statistical comparisons. <strong>c</strong>, PCA of ssGSEA-based analysis using ConsensusTME estimations (n = 38 samples from n = 8 independent patients). <strong>d</strong>, Principal-component feature loadings (magnitude and direction) of c. Vectors are colored according to cell type (<a href="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/tree/Figures/Figures/supplementary/3">Supplementary Fig. 3a</a>), for example, monocytes and macrophages (M0, M1, M2; orange), B cells and plasma cells (light blue), and CD8+ and cytotoxic cells (yellow). <strong>e</strong>, Differential expression analysis of tumors classified as having high and low tumor cellularity using WES-derived tumor cellularity (TITAN), with the median of the cohort as a cutoff. Patient dependence was used as a covariate (n = 36 samples with both WES and mRNA data from n = 8 independent patients). Limma-moderated two-sided t-statistics were calculated by empirical Bayes moderation. Benjamini–Hochberg (BH) was used for false discovery rate (FDR) correction. FDR < 0.05 was considered as the threshold for differential expression. <strong>f</strong>, GO enrichment analysis of 28 highly expressed genes showed significant enrichment in samples with low tumor cellularity. Two-sided Fisher’s exact test was used to calculate BH-FDR-corrected P values (<a href="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/tree/Tables/Tables/supplementary">Supplementary Table 3d</a>). <strong>g</strong>,<strong>h</strong>, ssGSEA analysis of differentially expressed genes using hallmarks (g) and ConsensusTME (h) NES. Gene sets on the x axes were ranked according to their NES (<a href="https://github.com/cansysbio/HGSOC_TME_Heterogeneity/tree/Tables/Tables/supplementary">Supplementary Table 3b, c</a>). High NES values reflect high tumor cellularity. Dashed red lines indicate the median and ±1.96 median absolute deviations (modified z-score). Kernel density plots of observed and fitted normal distribution are shown (right). No significant difference between the observed and fitted distribution was detected (Shapiro–Wilk test, D’Agostino’s K2 test and Anderson–Darling test for normality distribution). Significance levels: NS (not significant), P > 0.05; *P < 0.05, **P < 0.01, ***P < 0.001 and ****P <  0.0001.
</p>
