# Parses copy number data and folders generated by Snakemake TitanCNA.
# The output data is stored in files that are used elsewhere to generate figures.
#
# writes data/cna.csv file
# writes data/segs.RData and data/sets_singleClone.Rdata files

rm(list=ls())

library(data.table)
library(Homo.sapiens)  # hg19 genome annotation
library(sslist)
library(TitanCNA)
library(rlist)


# Entrez ID, HGNC gene symbol map
gene_map = as.data.frame(org.Hs.egSYMBOL)

setwd("/Users/koplev01/GoogleDrive/projects/cambridge/ovarianCancerHeterogeneityChemo/repo/HGSOC_TME_Heterogeneity/4")

source("lib/parse.R")

# Results directory from TitanCNA
results_dir = "/Users/koplev01/mnt/mmlab/projects/OVCT/titanCNA/TitanCNA/scripts/snakemake"

# Load TitanCNA results table
opt_clust = loadOptClust(file.path(results_dir, "results/titan/hmm/optimalClusterSolution.txt"))

# Load all optimal fits
d = loadTitanResultsEnv(opt_clust$path, results_dir)

# evaluate segments of copy-number calls from TitanCNA
segs = lapply(1:nrow(opt_clust), function(i) {
	message(i)
	s = outputTitanSegments(d[[i]]$results, id=opt_clust$id[i], d[[i]]$convergeParams)
	return(s)
})

# Save list of CNA segments
save(segs, file="data/segs.RData")

# Split elements by clone
segs_subclonal = list()
for (i in 1:length(segs)) {
	s = segs[[i]]
	for (j in 1:max(s$Clonal_Cluster, na.rm=TRUE)) {
		message(j)
		sub_s = s[s$Clonal_Cluster == j, ]
		segs_subclonal = list.append(segs_subclonal, sub_s)  # append
	}
}

# Format segments for use with mageGRangesFromDataFrame()
segs_format = lapply(segs, function(s) {
	s$Chromosome = paste0("chr", s$Chromosome)
	s$start = s$Start_Position.bp.
	s$end = s$End_Position.bp.
	return(s)
})



# Load fits with 1 clone specified
d_single = loadTitanResultsEnv(
	paste0("results/titan/hmm/titanCNA_ploidy2/", as.character(opt_clust$barcode), "_cluster1"),
	results_dir)

segs_single = lapply(1:length(d_single), function(i) {
	message(i)
	s = outputTitanSegments(d_single[[i]]$results, id=opt_clust$barcode[i], d_single[[i]]$convergeParams)
})

save(segs_single, file="data/segs_singleClone.RData")

# lapply(segs_single, function(x) unique(x$Clonal_Cluster))
# lapply(segs_single, function(x) nrow(x))
lapply(segs_single, function(x) table(x$Copy_Number))

# Evaluate copy numbers for all gene intervals

# Load hg19 gene annotations
gene_coords = genes(TxDb.Hsapiens.UCSC.hg19.knownGene)



cna_genes = lapply(1:length(segs_format), function(i) {
	# Extract genomic ranges from TitanCNA segments
	cna_coords = makeGRangesFromDataFrame(segs_format[[i]])

	# Annotate each gene range by CNA coordinate range
	merged_coords = mergeByOverlaps(gene_coords, cna_coords)

	# Append range info
	idx = match(merged_coords$cna_coords, cna_coords)
	merged_coords = cbind(merged_coords, segs_format[[i]][idx, ])

	return(merged_coords)
})


gene_ids = unique(unlist(lapply(cna_genes, function(x) x$gene_id)))


cna_mat = getFeatureMatrix(cna_genes, gene_ids, "Copy_Number")

# Combine counts and annotation into single table
annot_cna_mat = cbind(
	data.frame(
		entrez_id=rownames(cna_mat),
		hgnc_symbol=gene_map$symbol[match(rownames(cna_mat), gene_map$gene_id)]
	),
	cna_mat
)

write.csv(annot_cna_mat, file="data/cna.csv", row.names=FALSE)
